### PR TITLE
feat: have required keys for queries and matches and as instance attr

### PR DESCRIPTION
**Pull request by JoanFM**
**Changes introduced**
Allow the Ranker to have different set of keys extracted from Documents for Query and Matches

### CODE REVIEW

File: tests/unit/drivers/rank/aggregate/test_aggregate_matches_rank_driver.py [hunk 1]
```diff
@@ -27,8 +31,7 @@ def exec_fn(self):
 
 class MockLengthRanker(Chunk2DocRanker):
     def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.required_keys = {'length'}
+        super().__init__(query_required_keys=('length', ), match_required_keys=('length', ), *args, **kwargs)
```
Comment (cristianmtr) [inline tests/unit/drivers/rank/aggregate/test_aggregate_matches_rank_driver.py:34]: Maybe add a test with more than one required key?

Comment (JoanFM) [reply tests/unit/drivers/rank/aggregate/test_aggregate_matches_rank_driver.py:34]: any idea? the interpretability of this test would be lost

Comment (nan-wang) [reply tests/unit/drivers/rank/aggregate/test_aggregate_matches_rank_driver.py:34]: What about adding `discount` and `price`? We can set ranking score by `discount * price`.

```
```

**Review by nan-wang — CHANGES_REQUESTED** [review]
Can we make this backward compatible? There are pod images in the hub using `required_keys`. https://github.com/search?p=1&q=org%3Ajina-ai+required_keys&type=Code

File: tests/unit/drivers/rank/aggregate/test_aggregate_matches_rank_driver.py (hunk 1 above)
Comment (JoanFM) [reply tests/unit/drivers/rank/aggregate/test_aggregate_matches_rank_driver.py:34]: Great idea @nan-wang

**Comment by JoanFM** [comment]
> Can we make this backward compatible? There are pod images in the hub using `required_keys`. https://github.com/search?p=1&q=org%3Ajina-ai+required_keys&type=Code

Let me try to have it backward compatible

File: tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py [hunk 2]
```diff
@@ -48,21 +58,23 @@ def create_document_to_score():
...
+            match.id = str(match_id)
             parent_id = 10 * int(match_id)
-            match.parent_id = str(parent_id) * 8
+            match.parent_id = str(parent_id)
             match.length = int(match_id)
             # to be used by MaxRanker and MinRanker
             match.score.ref_id = chunk.id
             match.score.value = int(match_id)
+            match.tags['price'] = match.score.value
+            match.tags['discount'] = 0.5
```
Comment (nan-wang) [inline tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py:77]:
```suggestion
            match.tags['discount'] = self.discount
```

File: tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py [hunk 3]
```diff
@@ -21,13 +21,23 @@ def _get_score(self, match_idx, query_chunk_meta, match_chunk_meta, *args, **kwa
...
+        super().__init__(query_required_keys=('tags__price', 'tags__discount'), match_required_keys=('tags__price', 'tags__discount'), *args, **kwargs)
+
+    def _get_score(self, match_idx, query_chunk_meta, match_chunk_meta, *args, **kwargs):
+        price = match_chunk_meta[match_idx[0][self.COL_DOC_CHUNK_ID]]['tags__price']
+        discount = match_chunk_meta[match_idx[0][self.COL_DOC_CHUNK_ID]]['tags__discount']
+        return match_idx[0][self.COL_PARENT_ID], \
+               price - price * discount
+
+
 class SimpleChunk2DocRankDriver(Chunk2DocRankDriver):
```
Comment (nan-wang) [inline tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py:42]: discount = 0.5

File: tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py [hunk 4]
```diff
@@ -127,45 +139,25 @@ def create_chunk_chunk_matches_to_score():
...
-
 
-@pytest.mark.parametrize('keep_source_matches_as_chunks', [False, True])
-def test_chunk2doc_ranker_driver_max_ranker(keep_source_matches_as_chunks):
-    doc = create_document_to_score()
-    driver = SimpleChunk2DocRankDriver(keep_source_matches_as_chunks=keep_source_matches_as_chunks)
-    executor = MockMaxRanker()
-    driver.attach(executor=executor, runtime=None)
-    driver._traverse_apply(DocumentSet([doc, ]))
+    scale = 1 if not isinstance(executor, MockPriceDiscountRanker) else 0.5
```
Comment (nan-wang) [inline tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py:150]:
```suggestion
    scale = 1 if not isinstance(executor, MockPriceDiscountRanker) else MockPriceDiscountRanker.discount
```

**Review by nan-wang — CHANGES_REQUESTED** [review]
LGTM👍

File: tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py (hunk 2 above)
Comment (JoanFM) [reply tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py:77]: there is no self here right?

File: tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py (hunk 3 above)
Comment (JoanFM) [reply tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py:42]: But here I want to shpw that is passed to the Ranker

File: tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py (hunk 4 above)
Comment (JoanFM) [reply tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py:150]: but then the point of the test is lost I want to show how discount info arrives to the Ranker score function

File: tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py (hunk 3 above)
Comment (JoanFM) [reply tests/unit/drivers/rank/aggregate/test_chunk2doc_rank_drivers.py:42]: here the same

**Review by nan-wang — APPROVED** [review]
LGTM👍

**Pull request merged**